src/python/pants/core/util_rules/adhoc_binaries.py

```diff
@@ -0,0 +1,138 @@
+# Copyright 2023 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+from __future__ import annotations
+
+import os
+import sys
+from dataclasses import dataclass
+from textwrap import dedent  # noqa: PNT20
+
+from pants.core.util_rules.environments import EnvironmentTarget, LocalEnvironmentTarget
+from pants.core.util_rules.system_binaries import SEARCH_PATHS, TarBinary
+from pants.engine.fs import DownloadFile
+from pants.engine.internals.native_engine import EMPTY_DIGEST, Digest, FileDigest
+from pants.engine.internals.selectors import Get
+from pants.engine.platform import Platform
+from pants.engine.process import Process, ProcessResult
+from pants.engine.rules import collect_rules, rule
+from pants.util.frozendict import FrozenDict
+from pants.util.logging import LogLevel
+
+
+@dataclass(frozen=True)
+class PythonBuildStandaloneBinary:
+    """A Python interpreter for use by `@rule` code as an alternative to BashBinary scripts.
+
+    This interpreter is provided by Python Build Standalone https://gregoryszorc.com/docs/python-build-standalone/main/,
+    which has a few caveats. Namely it doesn't play nicely with third-party sdists. Meaning Pants'
+    scripts being run by PythonBuildStandalone should avoid third-party sdists.
+    """
+
+    SYMLINK_DIRNAME = ".python-build-standalone"
+
+    path: str
+    _digest: Digest
+
+    @property
+    def immutable_input_digests(self) -> FrozenDict[str, Digest]:
+        return FrozenDict({PythonBuildStandaloneBinary.SYMLINK_DIRNAME: self._digest})
+
+
+@rule(desc="Finding or downloading Python for scripts", level=LogLevel.TRACE)
+async def download_python_build_standalone(
+    platform: Platform, env_tgt: EnvironmentTarget, tar_binary: TarBinary
+) -> PythonBuildStandaloneBinary:
+    if env_tgt.val is None or isinstance(env_tgt.val, LocalEnvironmentTarget):
+        return PythonBuildStandaloneBinary(sys.executable, EMPTY_DIGEST)
+
+    url_plat, fingerprint, bytelen = {
+        "linux_arm64": (
+            "aarch64-unknown-linux-gnu",
+            "1ba520c0db431c84305677f56eb9a4254f5097430ed443e92fc8617f8fba973d",
+            23873387,
+        ),
+        "linux_x86_64": (
+            "x86_64-unknown-linux-gnu",
+            "7ba397787932393e65fc2fb9fcfabf54f2bb6751d5da2b45913cb25b2d493758",
+            26129729,
+        ),
+        "macos_arm64": (
+            "aarch64-apple-darwin",
+            "d732d212d42315ac27c6da3e0b69636737a8d72086c980daf844344c010cab80",
+            17084463,
+        ),
+        "macos_x86_64": (
+            "x86_64-apple-darwin",
+            "3948384af5e8d4ee7e5ccc648322b99c1c5cf4979954ed5e6b3382c69d6db71e",
+            17059474,
+        ),
```

Comment:
Can this be (lightly?) contorted to use an `ExternalTool` subsystem, so that it can be overridden if need be? Failing that, maybe this whole dict (and the template URL) could be supplied by options instead.

Reply:
We can't use `ExternalTool` due to a circular reference. I tried disentangling it and just gave up. Its possible but annoying.

We could have a subsystem, but why would we want to?

Taking a step back, this should be completely internal and shielded from the user. I don't think they can or should control our internal tools. What's your thinking here?

Reply:
> We could have a subsystem, but why would we want to?
> 
> Taking a step back, this should be completely internal and shielded from the user. I don't think they can or should control our internal tools. What's your thinking here?

People:
1. are behind firewalls, and might need to reconfigure things [á la](https://www.pantsbuild.org/docs/restricted-internet-access#setting-http_proxy-and-https_proxy)
2. attempt to cache all downloaded files to reduce flakiness (Dan, for example)
3. need to work around bugs in particular versions of things by upgrading to newer ones

Reply:
done